### ISSUE TITLE
invalid-index-returned false positive in stubs

### ISSUE THREAD

**Issue by jolaf** [PyCQA/pylint#4736]
The following file:
```python3
class A:
    def __index__(self) -> int: ...
```

produces the following output:

```python3
$ pylint A.pyi
************* Module A
A.pyi:2: [E0305(invalid-index-returned), A.__index__] __index__ does not return int
```

This looks like a false positive.

**Comment by OJFord** [comment]
Can't speak for the others, but the `__iter__` one seems to be a regression.

**Comment by jolaf** [comment]
It looks to me that all the return type checks should not be performed if method body is not defined or the method throws any unconditional exception.

**Comment by ntjess** [comment]
Is this related to https://github.com/microsoft/vscode-python/issues/16807, or should I open a new issue here for that problem?

**Comment by Pierre-Sassoulas** [comment]
@ntjess it look like it's related.

**Comment by jolaf** [comment]
Thanks @DanielNoord @cdce8p !!
